### PR TITLE
Standardizing makefile and adding automated builds

### PR DESCRIPTION
**Pull request by davidjumani**
*Description of changes:*
Standardizes the makefile as per other providers and adds automated builds via cloudbuild.yaml

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -1,10 +1,63 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ROOT_DIR_RELATIVE := .
+
+include $(ROOT_DIR_RELATIVE)/common.mk
```
Comment (maxdrib) [inline Makefile:17]: Why do we need this?

**Review by maxdrib — COMMENTED** [review]
Could you share an example that this refactoring was based off? 

Additionally, what value does the refactoring itself provide? As I understand, it is a fairly large change which interrupts some processes we have established as part of development and release

It might also be easier to review this Makefile change if we split it up into several PR's - maybe one for introducing the new variables and renaming old ones, and then another for the new targets? Otherwise this is a big change which is difficult to sign off on, especially when the value isn't immediately obvious.

**Comment by davidjumani** [comment]
@maxdrib It's based off other providers eg: aws, openstack
For most of it I've just offloaded the tasks to get the necessary binary tools to the hack/tools makefile
The rest is just moving things about and grouped them together

File: Makefile [hunk 2]
```diff
@@ -16,85 +69,138 @@ endif
...
-# file as xyz: ## something, and then pretty-format the target and help. Then,
-# if there's a line with ##@ something, that gets pretty-printed as a category.
-# More info on the usage of ANSI control characters for terminal formatting:
-# https://en.wikipedia.org/wiki/ANSI_escape_code#SGR_parameters
-# More info on the awk command:
-# http://linuxcommand.org/lc3_adv_awk.php
+.PHONY: binaries
+binaries: $(CONTROLLER_GEN KUSTOMIZE) $(GOLANGCI_LINT) $(STATIC_CHECK) $(GINKGO_V1) $(GINKGO_V2) $(MOCKGEN) $(KUSTOMIZE) managers # Builds and installs all binaries
 
-help: ## Display this help.
```
Comment (rejoshed) [inline Makefile:48]: Why are removing the help command and it's output?

File: Makefile [hunk 3]
```diff
@@ -16,85 +69,138 @@ endif
...
-	@awk 'BEGIN {FS = ":.*##"; printf "\nUsage:\n  make \033[36m<target>\033[0m\n"} /^[a-zA-Z_0-9-]+:.*?##/ { printf "  \033[36m%-15s\033[0m %s\n", $$1, $$2 } /^##@/ { printf "\n\033[1m%s\033[0m\n", substr($$0, 5) } ' $(MAKEFILE_LIST)
+.PHONY: managers
+managers:
+	$(MAKE) manager-cloudstack-infrastructure
 
- MANIFEST_GEN_INPUTS=$(shell find ./api ./controllers -type f -name "*test*" -prune -o -name "*zz_generated*" -prune -o -print)
```
Comment (rejoshed) [inline Makefile:51]: The original makefile from kubebuilder was really weak. It actually took a lot of work to make it build everything based on changed files.

I'm not at all for removing these features from the make file without something to replace them.

**Review by rejoshed — CHANGES_REQUESTED** [review]
> Standardizes the makefile as per other providers and adds automated builds via cloudbuild.yaml

I don't think we need to standardize everything. Particularly I don't want the help menu removed.

Otherwise, I will try out this branch shortly to check that the targets continue to work as expected.

File: Makefile (hunk 2 above)
Comment (davidjumani) [reply Makefile:48]: It is there in the common.mk
That way any file which needs a help can import it instead of adding it's own

```
[10:34:02] ~/lab/capi ➦ standardize-makefile $ make help

Usage:
  make <target>
  manager-cloudstack-infrastructure  Build manager binary.
  fmt              Run go fmt on the whole project.
  vet              Run go vet on the whole project.
  lint             Run linting for the project.
  modules          Runs go mod to ensure proper vendoring.
  generate-mocks   Generate mocks needed for testing. Primarily mocks of the cloud package.
  generate-deepcopy  Generate code containing DeepCopy, DeepCopyInto, and DeepCopyObject method implementations.
  generate-manifests  Generates crd, webhook, rbac, and other configuration manifests from kubebuilder instructions in go comments.
  build            Build manager binary.
  build-for-docker  Build manager binary for docker image building.
  run              Run a controller from your host.
  deploy           Deploy controller to the K8s cluster specified in ~/.kube/config.
  undeploy         Undeploy controller from the K8s cluster specified in ~/.kube/config.
  docker-build     Build docker image containing the controller manager.
  docker-push      Push docker image with the manager.
  tilt-up          Setup and run tilt for development.
  kind-cluster     Create a kind cluster with a local Docker repository.
  cluster-api      Clone cluster-api repository for tilt use.
  test             Run tests. At the moment this is only unit tests.
  clean            Cleans up everything.
  release-manifests  Create kustomized release manifest in $RELEASE_DIR (defaults to out).
  release-staging  Builds and push container images and manifests to the staging bucket.
  release-templates  Generate release templates
  upload-staging-artifacts  Upload release artifacts to the staging bucket
  help             Display this help

```

**Comment by davidjumani** [comment]
@rejoshed The code for the help menu is not removed, as shown. It's just moved to avoid duplication
I think it is a good idea to standardize. If anything it cleans up the make file and offloads downloading tools to the /hack/tools Makefile which handles managing any tool required to build capc, the same way it handles downloading binaries to create the documentation

File: Makefile [hunk 4]
```diff
@@ -107,122 +213,98 @@ docker-build: generate-deepcopy build-for-docker .dockerflag.mk ## Build docker
...
-	@# The following is a slightly funky way to make sure the ginkgo statements are removed regardless the test results.
-	@ginkgo_v2 --label-filter="!integ" --cover -coverprofile cover.out --covermode=atomic -v ./api/... ./controllers/... ./pkg/...; EXIT_STATUS=$$?;\
-		./hack/testing_ginkgo_recover_statements.sh --remove; exit $$EXIT_STATUS
-
-.PHONY: generate-mocks
-generate-mocks: bin/mockgen generate-deepcopy pkg/mocks/mock_client.go $(shell find ./pkg/mocks -type f -name "mock*.go") ## Generate mocks needed for testing. Primarily mocks of the cloud package.
-pkg/mocks/mock%.go: $(shell find ./pkg/cloud -type f -name "*test*" -prune -o -print)
-	go generate ./...
-
-##@ Tilt
```
[the remark this answers is not in the document]
Comment (davidjumani) [reply Makefile:193]: Added back section headers

File: Makefile (hunk 1 above)
Comment (davidjumani) [reply Makefile:17]: It contains common functionality. Eg: The help menu

**Comment by rohityadavcloud** [comment]
ping @davidjumani can you rebase/fix conflicts, and then @rejoshed can you do a round of review; thanks both.

**Comment by davidjumani** [comment]
Conflict fixed. Over to you @rejoshed

**Comment by rejoshed** [comment]
@davidjumani Thank you for your patience. I've been scrambling to get the multi-endpoint feature through in time to coordinate a release, and so I've fallen behind on PRs and general maintenance of this repo. I have just put through many of those changes and so hopefully will be getting caught up shortly.

I will prioritize this PR as next in line aside from anything release related that pulls me away.

File: Makefile [hunk 5]
```diff
@@ -112,122 +217,103 @@ docker-build: generate-deepcopy build-for-docker .dockerflag.mk ## Build docker
...
+tilt-up: cluster-api kind-cluster cluster-api/tilt-settings.json generate-manifests cloud-config ## Setup and run tilt for development.
 	export CLOUDSTACK_B64ENCODED_SECRET=$$(base64 -w0 -i cloud-config 2>/dev/null || base64 -b 0 -i cloud-config) && cd cluster-api && tilt up
 
 .PHONY: kind-cluster
 kind-cluster: cluster-api ## Create a kind cluster with a local Docker repository.
 	-./cluster-api/hack/kind-install-for-capd.sh
 
 cluster-api: ## Clone cluster-api repository for tilt use.
-	git clone --branch v1.0.0 https://github.com/kubernetes-sigs/cluster-api.git
+	git clone --branch v1.0.0 --depth 1 https://github.com/kubernetes-sigs/cluster-api.git
```
Comment (rejoshed) [inline Makefile:234]: Nice little addition : )

File: Makefile (hunk 1 above)
Comment (rejoshed) [reply Makefile:17]: 1). Could we use `REPO_ROOT` here instead of `ROOT_DIR_RELATIVE`?

2). In changing to use the more standard REPO_ROOT, what is the reason to change to the `git rev-parse` method of finding the root directory?
` export REPO_ROOT := $(shell git rev-parse --show-toplevel)`

We used to use:
`export PROJECT_DIR := $(shell dirname $(realpath $(lastword $(MAKEFILE_LIST))))`
which utilizes the Makefile's own location and the `make` command's built in list.

And it looks like VSphere and Azure Use:
`ROOT_DIR:=$(shell dirname $(realpath $(firstword $(MAKEFILE_LIST))))`

Comment (rejoshed) [reply Makefile:17]: > It contains common functionality. Eg: The help menu

Yeesh... Our Makefile's grown that much hasn't it. Haha

File: Makefile [hunk 6]
```diff
@@ -1,10 +1,63 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+ROOT_DIR_RELATIVE := .
+
+include $(ROOT_DIR_RELATIVE)/common.mk
+
+# Directories
+TOOLS_DIR := hack/tools
```
Comment (rejoshed) [inline Makefile:20]: Let's make this variable repo relative to begin with.

So, `$(REPO_ROOT)/hack/tools`

I see later in the file we prepend the root dir anyway, so this would just be a safer way to specify this directory.

I've not search all throughout yet or anything, but I think we should do the same with other Path vars if not already absolute.

File: Makefile (hunk 2 above)
Comment (rejoshed) [reply Makefile:48]: Thank you for cleaning up the help menu and targets/descriptions.

**Review by rejoshed — CHANGES_REQUESTED** [review]
Overall I really like the direction of this PR, but I've left some comments on small things, and as of now, on `main` the target `make test` works for me locally, but on the `automated-builds` branch I get the following issue.

``` sh
$ make modules
go mod tidy -compat=1.17
go: finding module for package sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2
sigs.k8s.io/cluster-api-provider-cloudstack/pkg/mocks imports
	sigs.k8s.io/cluster-api-provider-cloudstack/api/v1beta2: no matching versions for query "latest"
make: *** [modules] Error 1
```

I'm not yet sure why, but I'm getting this particular issue for `make test` and `make modules`. I'll try to dig into that a bit.

May just be something changed in main's go.mod?

File: hack/tools/Makefile [hunk 7]
```diff
@@ -67,6 +67,22 @@ CONVERSION_GEN := $(BIN_DIR)/conversion-gen
 $(CONVERSION_GEN): $(BIN_DIR) go.mod go.sum
 	go build -tags=tools -o $@ k8s.io/code-generator/cmd/conversion-gen
 
+GOLANGCI_LINT := $(BIN_DIR)/golangci-lint
```
Comment (rejoshed) [inline hack/tools/Makefile:70]: I really like moving the tools out of ./bin and into their own tool directory.

File: go.mod [hunk 8]
```diff
@@ -7,6 +7,7 @@ require (
 	github.com/go-logr/logr v1.2.3
 	github.com/golang/mock v1.6.0
 	github.com/hashicorp/go-multierror v1.1.1
+	github.com/onsi/ginkgo v1.16.5
```
Comment (rejoshed) [inline go.mod:10]: I'm not too sure why this change is needed. I thought all of our unit/controller tests have moved to Ginkgo V2, and the E2E tests are in their own module.

File: hack/tools/tools.go [hunk 9]
```diff
@@ -21,7 +21,11 @@ package tools
 import (
 	_ "github.com/a8m/envsubst"
```
Comment (rejoshed) [inline hack/tools/tools.go:22]: This is a very interesting way to vendor tools. Seems to work, kinda like it.

File: Makefile [hunk 10]
```diff
@@ -112,122 +217,103 @@ docker-build: generate-deepcopy build-for-docker .dockerflag.mk ## Build docker
...
 cluster-api/tilt-settings.json: hack/tilt-settings.json cluster-api
 	cp ./hack/tilt-settings.json cluster-api
 
-##@ End-to-End Testing
+##@ Tests
+## --------------------------------------
+## Tests
+## --------------------------------------
+
+export KUBEBUILDER_ASSETS=$(REPO_ROOT)/$(TOOLS_BIN_DIR)
```
Comment (rejoshed) [inline Makefile:244]: If we make `TOOLS_BIN_DIR` project root relative to begin with then this var can go away.

File: Makefile [hunk 11]
```diff
@@ -112,122 +217,103 @@ docker-build: generate-deepcopy build-for-docker .dockerflag.mk ## Build docker
...
 	kind delete clusters capi-test
+
+##@ Cleanup
+## --------------------------------------
+## Cleanup
+## --------------------------------------
+
+.PHONY: clean
+clean: ## Cleans up everything.
+	rm -rf $(RELEASE_DIR)
```
Comment (rejoshed) [inline Makefile:277]: Main has the removal of generated test templates. Will probably come in with a merge.

File: go.mod (hunk 8 above)
Comment (jweite-amazon) [reply go.mod:10]: We want to be very conservative about introduction of new packages into CAPC itself.  These have potential to introduce a release-blocking security issue.

Comment (rejoshed) [reply go.mod:10]: ++

File: Makefile (hunk 10 above)
Comment (davidjumani) [reply Makefile:252]: Not necessarily, it requires the absolute path not a relative one

**Comment by davidjumani** [comment]
> `make test`

Works fine now
